docs/remote-attestation.md

```diff
@@ -1,212 +1,111 @@
+activate U
+U->>C: GetPublicKeyResponse
+deactivate U
+C-->>C: Verify Enclave Public Key attestation
+
+Note over C,S: Send Encrypted Data
+
+loop
+   C-->>C: encrypt request
+   C->>U: InvokeRequest
+   activate U
+   U->>T: InvokeRequest
+   activate T
+   T-->>T: decrypt request
+   T-->>T: process request
+   T-->>T: encrypt response
+   T->>U: InvokeResponse
+   deactivate T
+   U->>C: InvokeResponse
+   deactivate U
+   C-->>C: decrypt response
+end
+```
 
-  Trusted Runtime-->>Trusted Runtime: Generate a Signing key pair
-  Trusted Runtime-->>Trusted Runtime: Put Signing public key hash in the AttestationReport request
-  Trusted Runtime->>TEE Platform: AttestationReport request
-  TEE Platform-->>TEE Platform: Generate AttestationReport
-  TEE Platform->>Trusted Runtime: Signed AttestationReport
+### Fetch Enclave Public Key
 
-  Note over Untrusted Launcher,Trusted Runtime: Initialization
+Initially, the client connects to the untrusted launcher and requests the
+attested enclave public key. This key was generated in the enclave, but it is
+then kept in memory by the corresponding untrusted launcher, so it may be served
+to the client without having to reach into the enclave every time.
 
-  Untrusted Launcher->>Trusted Runtime: Configuration and business logic
+The client then checks:
 
-  Client-->>Client: Generate a Signing key pair
+- that the attestation report is signed by the TEE manufacturer (e.g. AMD,
+  Intel)
+- that the attestation report additional data includes the hash of the public
+  key, to confirm that the key pair was in fact generated from inside the
+  enclave
+- that the attestation report measurement corresponds to a trusted version of
+  the enclave binary
 
-  Note over Client,Trusted Runtime: Handshake Start
+If any of these checks fails, the client refuses to go ahead.
 
-  Client-->>Client: Generate an Ephemeral key pair
-  Client->>Untrusted Launcher: ClientHello
-  Untrusted Launcher->>Trusted Runtime: ClientHello
-  Trusted Runtime-->>Trusted Runtime: Generate an Ephemeral key pair
+### Send Encrypted Data
 
-  Trusted Runtime->>Untrusted Launcher: ServerIdentity
-  Untrusted Launcher->>Client: ServerIdentity
-  Client-->>Client: Validate ServerIdentity
+The client also generates an encryption key pair ahead of time (it may be reused
+across invocations, or it may be generated for each invocation).
 
-  Client->>Untrusted Launcher: ClientIdentity
-  Untrusted Launcher->>Trusted Runtime: ClientIdentity
+For each invocation (consisting of a request followed by a response) the client
+concatenates its public key with the request payload, and encrypts the resulting
+blob with the enclave public key using
+[Hybrid Encryption](https://developers.google.com/tink/hybrid).
 
-  Client-->>Client: Derive Session Keys
-  Trusted Runtime-->>Trusted Runtime: Derive Session Keys
-  Note over Client,Trusted Runtime: Handshake Completed
+It then sends the encrypted message to the server, which ensures it gets
```

Comment:
I am not sure why the encrypted message ensures that the trusted enclave which generated the key gets the message. I think this is the only enclave which can decrypt it, no? But it could be send to a different one.

Reply:
I meant that the server ensures it, reworded it now.